File: lib/shire/projects.ex
defmodule Shire.Projects do
  import Ecto.Query
  alias Ecto.Multi
  alias Shire.Repo
  alias Shire.Projects.Project

  def create_project(name) do
    %Project{}
    |> Project.changeset(%{name: name})
    |> Repo.insert()
  end

  def get_project!(id), do: Repo.get!(Project, id)

  def get_project_by_id(id) do
    Repo.get(Project, id)
  end

  def get_project_by_name(name) do
    Repo.get_by(Project, name: name)
  end

  def get_project_by_name!(name) do
    Repo.get_by!(Project, name: name)
  end

  def list_projects do
    Repo.all(from(p in Project, order_by: [asc: p.name]))
  end

  def rename_project(%Project{} = project, new_name) do
    Multi.new()
    |> Multi.update(:project, Project.changeset(project, %{name: new_name}))
    |> Repo.transaction()
    |> case do
      {:ok, %{project: project}} -> {:ok, project}
      {:error, :project, changeset, _} -> {:error, changeset}
    end
  end

  def delete_project(%Project{} = project) do
    Repo.delete(project)
  end
end

File: src/services/projects.ts
import { eq } from "drizzle-orm";
import { getDb, schema } from "../db";

const { projects } = schema;

export function listProjects() {
  return getDb().select().from(projects).orderBy(projects.name).all();
}

export function getProject(id: string) {
  return getDb().select().from(projects).where(eq(projects.id, id)).get();
}

export function getProjectByName(name: string) {
  return getDb().select().from(projects).where(eq(projects.name, name)).get();
}

export function createProject(name: string) {
  return getDb().insert(projects).values({ name }).returning().get();
}

export function renameProject(id: string, name: string) {
  return getDb()
    .update(projects)
    .set({ name, updatedAt: new Date().toISOString() })
    .where(eq(projects.id, id))
    .returning()
    .get();
}

export function deleteProject(id: string) {
  return getDb().delete(projects).where(eq(projects.id, id)).returning().get();
}
